src/vm/functions/principals.rs

```diff
@@ -38,13 +53,97 @@ pub fn special_is_standard(
         _ => return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, owner).into()),
     };
 
-    let address_is_mainnet = version == C32_ADDRESS_VERSION_MAINNET_MULTISIG
-        || version == C32_ADDRESS_VERSION_MAINNET_SINGLESIG;
-    let address_is_testnet = version == C32_ADDRESS_VERSION_TESTNET_MULTISIG
-        || version == C32_ADDRESS_VERSION_TESTNET_SINGLESIG;
+    let address_is_mainnet = version_matches_mainnet(version);
+    let address_is_testnet = version_matches_testnet(version);
     let context_is_mainnet = env.global_context.mainnet;
 
     Ok(Value::Bool(
         (address_is_mainnet && context_is_mainnet) || (address_is_testnet && !context_is_mainnet),
     ))
 }
+
+pub fn native_principal_parse(principal: Value) -> Result<Value> {
+    let (version_byte, hash_bytes) = match principal {
+        Value::Principal(PrincipalData::Standard(StandardPrincipalData(version, bytes))) => {
+            (version, bytes)
+        }
+        Value::Principal(PrincipalData::Contract(QualifiedContractIdentifier { issuer, name })) => {
+            (issuer.0, issuer.1)
+        }
+        _ => {
+            return Err(CheckErrors::TypeValueError(TypeSignature::PrincipalType, principal).into())
+        }
+    };
+
+    if !version_matches_mainnet(version_byte) && !version_matches_testnet(version_byte) {
+        return Err(CheckErrors::InvalidVersionByte.into());
+    }
+
+    let buffer_data = match Value::buff_from(hash_bytes.to_vec()) {
+        Ok(data) => data,
+        Err(err) => return Err(err),
+    };
+
+    let tuple_data = TupleData::from_data(vec![
+        (
+            ClarityName::try_from("version".to_owned()).unwrap(),
+            Value::buff_from_byte(version_byte),
+        ),
+        (
+            ClarityName::try_from("hash-bytes".to_owned()).unwrap(),
+            buffer_data,
+        ),
+    ])?;
+
+    Ok(Value::Tuple(tuple_data))
+}
+
+pub fn native_principal_construct(version: Value, hash_bytes: Value) -> Result<Value> {
+    // Check the version byte.
+    let verified_version = match version {
+        Value::Sequence(SequenceData::Buffer(BuffData { ref data })) => data,
+        _ => return Err(CheckErrors::TypeValueError(TypeSignature::UIntType, version).into()),
+    };
+
+    if verified_version.len() != 1 {
+        return Err(CheckErrors::TypeValueError(
+            TypeSignature::SequenceType(SequenceSubtype::BufferType(BufferLength(1))),
+            version,
+        )
+        .into());
+    }
```

Comment:
Same as above -- I think this should be a runtime error, if I pass this method a variable for the version byte, and that variable holds a 0-length buff, the type checker will accept the input (because the type system only check max-length), and this case will fail.